client/client.go

```diff
@@ -197,7 +197,8 @@ type Client struct {
 	// invalidAllocs is a map that tracks allocations that failed because
 	// the client couldn't initialize alloc or task runners for it. This can
 	// happen due to driver errors
-	invalidAllocs map[string]struct{}
+	invalidAllocs     map[string]struct{}
+	invalidAllocsLock sync.Mutex
```

Comment:
Do you have the race detector output for this handy? Skimming some from `go test -race` in `client` it looks like only tests access this field concurrently. I'd rather fix the tests then introduce a lock that's unnecessary at runtime. During implementation of this field we tried to be careful that only the `Client.run` goroutine accessed it.

Reply:
I don't have race detector output for regular run of application, but I see accesses in two goroutines:

1. In `watchAllocations()` goroutine, we check for invalid alloc by accessing map directly [1].  `watchAllocations` goroutine is started `run()`[2]
2. In `run()` goroutine, we access the field through some intermediate functions, in `runAllocs()->handleInvalidAllocs`[3] and `runAllocs()->removeAlloc()`[4]

[1] https://github.com/hashicorp/nomad/blob/d1f12fd3cb0805b5b7d75d0f779cf5f5369712bc/client/client.go#L1867
[2] https://github.com/hashicorp/nomad/blob/d1f12fd3cb0805b5b7d75d0f779cf5f5369712bc/client/client.go#L1471
[3] https://github.com/hashicorp/nomad/blob/d1f12fd3cb0805b5b7d75d0f779cf5f5369712bc/client/client.go#L2042
[4] https://github.com/hashicorp/nomad/blob/d1f12fd3cb0805b5b7d75d0f779cf5f5369712bc/client/client.go#L2114-L2116